testutils/testcluster: TestManualReplication failed

testutils/testcluster.TestManualReplication [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3797088&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3797088&tab=artifacts#/) on master @ [38fd4064f40c574d0b2aee02f414787a2760fcec](https://github.com/cockroachdb/cockroach/commits/38fd4064f40c574d0b2aee02f414787a2760fcec):

Fatal error:

```
panic: assignment to entry in nil map

```
Stack: 

```
goroutine 117375 [running]:
github.com/cockroachdb/pebble.(*tableCacheShard).addNode(0xc00156ac80, 0xc005cc8180, 0xc001098908)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/table_cache.go:551 +0x85
github.com/cockroachdb/pebble.(*tableCacheShard).findNode(0xc00156ac80, 0xc00572e460, 0xc001098908)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/table_cache.go:487 +0x2c6
github.com/cockroachdb/pebble.(*tableCacheShard).newIters(0xc00156ac80, 0xc0058ae760, 0xc0041cea30, 0x0, 0xc001098908)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/table_cache.go:306 +0x85
github.com/cockroachdb/pebble.(*tableCacheContainer).newIters(0x2b, 0xc0058ae7d8, 0x106b6d0, 0xc0027527e0)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/table_cache.go:116 +0x45
github.com/cockroachdb/pebble.(*levelIter).loadFile(0xc0041ce9e0, 0xc00572e460, 0x1)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/level_iter.go:339 +0x20e
github.com/cockroachdb/pebble.(*levelIter).SeekPrefixGE(0xc0041ce9e0, {0xc0014602d0, 0xb, 0x44}, {0xc002c38a80, 0xb, 0x2b}, 0x0)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/level_iter.go:409 +0xbf
github.com/cockroachdb/pebble.(*mergingIter).seekGE(0xc0041ce238, {0xc002c38a80, 0x0, 0x0}, 0x0, 0xb)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/merging_iter.go:830 +0xf1
github.com/cockroachdb/pebble.(*mergingIter).SeekPrefixGE(0xc0041ce238, {0xc0014602d0, 0x0, 0x2b}, {0xc002c38a80, 0xa, 0xa}, 0x0)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/merging_iter.go:899 +0x96
github.com/cockroachdb/pebble.(*Iterator).SeekPrefixGE(0xc0041ce000, {0xc002c38a80, 0xb, 0x2b})
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/iterator.go:776 +0x48d
github.com/cockroachdb/cockroach/pkg/storage.(*pebbleIterator).SeekGE(0xc0030bafc0, {{0xc004acd430, 0xa, 0xa}, {0x0, 0x0, 0x0}})
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/pebble_iterator.go:263 +0xb7
github.com/cockroachdb/cockroach/pkg/storage.(*pebbleMVCCScanner).get(0xc0027cc700, {0x50f17d8, 0xc0024b84b0})
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/pebble_mvcc_scanner.go:226 +0x88
github.com/cockroachdb/cockroach/pkg/storage.mvccGet({0x50f17d8, 0xc0024b84b0}, {0x5278dd0, 0xc0030bafc0}, {0xc004acd430, 0x0, 0xa}, {0x0, 0x0, 0x0}, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/mvcc.go:898 +0x416
github.com/cockroachdb/cockroach/pkg/storage.MVCCGet({0x50f17d8, 0xc0024b84b0}, {0x518c9a0, 0xc0013499e0}, {0xc004acd430, 0xa, 0xa}, {0x0, 0x0, 0x0}, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/mvcc.go:858 +0x1ee
github.com/cockroachdb/cockroach/pkg/server.(*nodeTombstoneStorage).IsDecommissioned(0xc000c44740, {0x50f17d8, 0xc0024b84b0}, 0x2)
	/go/src/github.com/cockroachdb/cockroach/pkg/server/node_tombstone_storage.go:62 +0x3d7
github.com/cockroachdb/cockroach/pkg/server.NewServer.func1({0x50f17d8, 0xc0024b84b0}, 0x1, 0x2)
	/go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:308 +0x71
github.com/cockroachdb/cockroach/pkg/server.NewServer.func3(0x50f17d8)
	/go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:342 +0x31
github.com/cockroachdb/cockroach/pkg/rpc.(*HeartbeatService).Ping(0xc00073d680, {0x50f17d8, 0xc004655440}, 0xc0035f3860)
	/go/src/github.com/cockroachdb/cockroach/pkg/rpc/heartbeat.go:172 +0x498
github.com/cockroachdb/cockroach/pkg/rpc._Heartbeat_Ping_Handler.func1({0x50f17d8, 0xc004655440}, {0x3b5a6a0, 0xc0035f3860})
	/go/src/github.com/cockroachdb/cockroach/pkg/rpc/heartbeat.pb.go:283 +0x78
github.com/cockroachdb/cockroach/pkg/util/tracing.ServerInterceptor.func1({0x50f17d8, 0xc004655440}, {0x3b5a6a0, 0xc0035f3860}, 0xc00672e680, 0xc00394e2d0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/grpc_interceptor.go:115 +0x234
google.golang.org/grpc.chainUnaryInterceptors.func1.1({0x50f17d8, 0xc004655440}, {0x3b5a6a0, 0xc0035f3860})
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1131 +0x63
github.com/cockroachdb/cockroach/pkg/rpc.NewServer.func1({0x50f17d8, 0xc004655440}, {0x3b5a6a0, 0xc0035f3860}, 0x3, 0xc0068dd900)
	/go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:183 +0x83
google.golang.org/grpc.chainUnaryInterceptors.func1.1({0x50f17d8, 0xc004655440}, {0x3b5a6a0, 0xc0035f3860})
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1134 +0x87
github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.unaryInterceptor({{{0x10000c0042c2b60}}}, {0x50f17d8, 0xc004655440}, {0x3b5a6a0, 0xc0035f3860}, 0xc00672e680, 0xc0068dd900)
	/go/src/github.com/cockroachdb/cockroach/pkg/rpc/auth.go:71 +0x173
google.golang.org/grpc.chainUnaryInterceptors.func1.1({0x50f17d8, 0xc004655440}, {0x3b5a6a0, 0xc0035f3860})
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1134 +0x87
google.golang.org/grpc.chainUnaryInterceptors.func1({0x50f17d8, 0xc004655440}, {0x3b5a6a0, 0xc0035f3860}, 0xc00672e680, 0xc00394e2d0)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1136 +0x159
github.com/cockroachdb/cockroach/pkg/rpc._Heartbeat_Ping_Handler({0x3762d00, 0xc00073d680}, {0x50f17d8, 0xc004655440}, 0xc0035f3800, 0xc00156d8c0)
	/go/src/github.com/cockroachdb/cockroach/pkg/rpc/heartbeat.pb.go:285 +0x138
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0020fc000, {0x5166d40, 0xc000764600}, 0xc003950000, 0xc0022d70b0, 0x7242210, 0x0)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1297 +0xccf
google.golang.org/grpc.(*Server).handleStream(0xc0020fc000, {0x5166d40, 0xc000764600}, 0xc003950000, 0x0)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1626 +0xa2a
google.golang.org/grpc.(*Server).serveStreams.func1.2()
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:941 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:939 +0x294
```
<details><summary>Log preceding fatal error</summary>
<p>

```
I211125 10:44:36.467823 113102 3@vendor/github.com/cockroachdb/pebble/event.go:612  [n3,pebble,s3] 243  [JOB 9] compacting(elision-only) L6 [000004] (1.3 K) + L6 [] (0 B)
I211125 10:44:36.467966 113102 3@vendor/github.com/cockroachdb/pebble/event.go:644  [n3,pebble,s3] 244  [JOB 9] compacting: sstable created 000013
I211125 10:44:36.468151 113102 3@vendor/github.com/cockroachdb/pebble/event.go:616  [n3,pebble,s3] 245  [JOB 9] compacted(elision-only) L6 [000004] (1.3 K) + L6 [] (0 B) -> L6 [000013] (1016 B), in 0.0s, output rate 4.3 M/s
I211125 10:44:36.468271 113144 3@vendor/github.com/cockroachdb/pebble/event.go:612  [n3,pebble,s3] 246  [JOB 10] compacting(elision-only) L6 [000006] (1.0 K) + L6 [] (0 B)
I211125 10:44:36.468422 113144 3@vendor/github.com/cockroachdb/pebble/event.go:616  [n3,pebble,s3] 247  [JOB 10] compacted(elision-only) L6 [000006] (1.0 K) + L6 [] (0 B) -> L6 [] (0 B), in 0.0s, output rate 0 B/s
I211125 10:44:36.468545 113204 3@vendor/github.com/cockroachdb/pebble/event.go:612  [n3,pebble,s3] 248  [JOB 11] compacting(elision-only) L6 [000007] (1.0 K) + L6 [] (0 B)
I211125 10:44:36.468685 113204 3@vendor/github.com/cockroachdb/pebble/event.go:616  [n3,pebble,s3] 249  [JOB 11] compacted(elision-only) L6 [000007] (1.0 K) + L6 [] (0 B) -> L6 [] (0 B), in 0.0s, output rate 0 B/s
I211125 10:44:36.468801 113216 3@vendor/github.com/cockroachdb/pebble/event.go:612  [n3,pebble,s3] 250  [JOB 12] compacting(elision-only) L6 [000008] (1.2 K) + L6 [] (0 B)
I211125 10:44:36.468973 113216 3@vendor/github.com/cockroachdb/pebble/event.go:644  [n3,pebble,s3] 251  [JOB 12] compacting: sstable created 000014
I211125 10:44:36.468960 113203 3@vendor/github.com/cockroachdb/pebble/event.go:648  [n3,pebble,s3] 252  [JOB 10] sstable deleted 000006
I211125 10:44:36.468985 113143 3@vendor/github.com/cockroachdb/pebble/event.go:648  [n3,pebble,s3] 253  [JOB 9] sstable deleted 000004
I211125 10:44:36.469060 113215 3@vendor/github.com/cockroachdb/pebble/event.go:648  [n3,pebble,s3] 254  [JOB 11] sstable deleted 000007
I211125 10:44:36.469219 113216 3@vendor/github.com/cockroachdb/pebble/event.go:616  [n3,pebble,s3] 255  [JOB 12] compacted(elision-only) L6 [000008] (1.2 K) + L6 [] (0 B) -> L6 [000014] (931 B), in 0.0s, output rate 3.0 M/s
I211125 10:44:36.469352 113305 3@vendor/github.com/cockroachdb/pebble/event.go:648  [n3,pebble,s3] 256  [JOB 12] sstable deleted 000008
I211125 10:44:36.469423 8530 kv/kvserver/replica_command.go:2128  [n1,s1,r44/1:/{Table/53-Max}] 257  change replicas (add [(n2,s2):2VOTER_INCOMING (n3,s3):3VOTER_INCOMING] remove []): existing descriptor r44:/{Table/53-Max} [(n1,s1):1, (n2,s2):2LEARNER, (n3,s3):3LEARNER, next=4, gen=3, sticky=9223372036.854775807,2147483647]
I211125 10:44:36.489425 8530 kv/kvserver/replica_raft.go:298  [n1,s1,r44/1:/{Table/53-Max}] 258  proposing ENTER_JOINT(v2 v3) [(n2,s2):2VOTER_INCOMING (n3,s3):3VOTER_INCOMING]: after=[(n1,s1):1 (n2,s2):2VOTER_INCOMING (n3,s3):3VOTER_INCOMING] next=4
I211125 10:44:36.491995 8530 kv/kvserver/replica_command.go:2128  [n1,s1,r44/1:/{Table/53-Max}] 259  change replicas (add [] remove []): existing descriptor r44:/{Table/53-Max} [(n1,s1):1, (n2,s2):2VOTER_INCOMING, (n3,s3):3VOTER_INCOMING, next=4, gen=4, sticky=9223372036.854775807,2147483647]
I211125 10:44:36.509149 8530 kv/kvserver/replica_raft.go:298  [n1,s1,r44/1:/{Table/53-Max}] 260  proposing LEAVE_JOINT: after=[(n1,s1):1 (n2,s2):2 (n3,s3):3] next=4
I211125 10:44:36.511857 116204 2@rpc/context.go:1045  [n1] 261  dialing n3: 127.0.0.1:37899 (system)
I211125 10:44:36.511798 116175 2@rpc/context.go:1045  [n1] 262  dialing n2: 127.0.0.1:37127 (system)
I211125 10:44:36.517129 117230 2@rpc/context.go:1045  [n2] 263  dialing n3: 127.0.0.1:37899 (default)
I211125 10:44:36.517268 8530 testutils/testcluster/testcluster.go:106  [-] 264  TestCluster quiescing nodes
W211125 10:44:36.517767 115963 kv/kvserver/intentresolver/intent_resolver.go:578  [-] 265  failed to cleanup transaction intents: failed to resolve intents: kv/kvserver/intentresolver/intent_resolver.go:901: stopping
W211125 10:44:36.517905 81696 jobs/registry.go:851  [n3] 266  canceling all adopted jobs due to stopper quiescing
I211125 10:44:36.518038 111958 2@rpc/context.go:688  [n1] 267  closing [{n3: 127.0.0.1:37899 (default)} {n0: 127.0.0.1:37899 (default)}]
I211125 10:44:36.518081 76311 2@rpc/context.go:688  [n3] 268  closing [{n1: 127.0.0.1:45139 (default)} {n0: 127.0.0.1:45139 (default)}]
I211125 10:44:36.518140 73764 2@rpc/context.go:688  [n3] 269  closing [{n0: 127.0.0.1:45139 (system)}]
I211125 10:44:36.518164 75964 2@rpc/context.go:688  [n3] 270  closing [{n1: 127.0.0.1:45139 (system)}]
I211125 10:44:36.518427 89689 2@rpc/context.go:688  [n1] 271  closing [{n1: 127.0.0.1:45139 (default)} {n0: 127.0.0.1:45139 (default)}]
W211125 10:44:36.518496 28603 jobs/registry.go:851  [n1] 272  canceling all adopted jobs due to stopper quiescing
W211125 10:44:36.518689 116204 2@rpc/nodedialer/nodedialer.go:198  [n1] 273  unable to connect to n3: failed to connect to n3 at 127.0.0.1:37899: stopped
I211125 10:44:36.518248 110711 2@rpc/context.go:688  [n1] 274  closing [{n2: 127.0.0.1:37127 (default)} {n0: 127.0.0.1:37127 (default)}]
W211125 10:44:36.518963 76945 sql/sqlliveness/slinstance/slinstance.go:220  [n3] 275  exiting heartbeat loop
W211125 10:44:36.518991 20919 sql/sqlliveness/slinstance/slinstance.go:220  [n1] 276  exiting heartbeat loop
W211125 10:44:36.518737 116175 2@rpc/nodedialer/nodedialer.go:198  [n1] 277  unable to connect to n2: failed to connect to n2 at 127.0.0.1:37127: stopped
I211125 10:44:36.518969 9124 2@rpc/context.go:688  [n1] 279  closing [{n0: 127.0.0.1:37899 (system)}]
W211125 10:44:36.519110 117230 2@rpc/nodedialer/nodedialer.go:198  [n2] 278  unable to connect to n3: failed to connect to n3 at 127.0.0.1:37899: stopped
W211125 10:44:36.519266 67986 jobs/registry.go:851  [n2] 280  canceling all adopted jobs due to stopper quiescing
I211125 10:44:36.519368 62633 2@rpc/context.go:688  [n2] 281  closing [{n1: 127.0.0.1:45139 (default)} {n0: 127.0.0.1:45139 (default)}]
I211125 10:44:36.519432 62263 2@rpc/context.go:688  [n2] 282  closing [{n1: 127.0.0.1:45139 (system)}]
I211125 10:44:36.519734 59966 2@rpc/context.go:688  [n2] 283  closing [{n0: 127.0.0.1:45139 (system)}]
E211125 10:44:36.517703 116649 kv/kvserver/consistency_queue.go:187  [n1,consistencyChecker,s1,r3/1:/System/{NodeLive…-tsd}] 284  result is ambiguous (server shutdown)
E211125 10:44:36.520009 116649 kv/kvserver/queue.go:1093  [n1,consistencyChecker,s1,r3/1:/System/{NodeLive…-tsd}] 285  result is ambiguous (server shutdown)
I211125 10:44:36.520399 116233 2@rpc/context.go:688  [n1] 286  closing [{n3: 127.0.0.1:37899 (system)} {n0: 127.0.0.1:37899 (system)}]
I211125 10:44:36.520481 116292 2@rpc/context.go:688  [n1] 287  closing [{n2: 127.0.0.1:37127 (system)} {n0: 127.0.0.1:37127 (system)}]
W211125 10:44:36.527180 63384 sql/sqlliveness/slinstance/slinstance.go:220  [n2] 288  exiting heartbeat loop
I211125 10:44:36.519185 9124 2@rpc/context.go:688  [n1] 289  closing [{n2: 127.0.0.1:37127 (system)}]
I211125 10:44:36.536359 11695 2@rpc/context.go:688  [n2] 290  closing [{n3: 127.0.0.1:37899 (default)}]
I211125 10:44:36.536462 11695 2@rpc/context.go:688  [n2] 291  closing [{n0: 127.0.0.1:37899 (default)}]
I211125 10:44:36.536537 117279 2@rpc/context.go:688  [n2] 292  closing [{n3: 127.0.0.1:37899 (default)} {n0: 127.0.0.1:37899 (default)}]
```
</p>
</details>
<details><summary>Help</summary>
<p>
See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)Parameters in this failure:

- TAGS=deadlock

- GOFLAGS=-parallel=4
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #62945 testutils/testcluster: TestManualReplication failed [C-test-failure O-robot branch-release-21.1]
</p>
</details>
/cc @cockroachdb/test-eng
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestManualReplication.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/closedts/sidetransport/sender_test.go
+++ b/pkg/kv/kvserver/closedts/sidetransport/sender_test.go
@@ -309,7 +309,13 @@ func (s *sideTransportGRPCServer) addr() net.Addr {
 
 func newMockSideTransportGRPCServer(stopper *stop.Stopper) (*sideTransportGRPCServer, error) {
 	receiver := newMockReceiver()
-	stopper.AddCloser(receiver)
+	if err := stopper.RunAsyncTask(context.Background(), "stopper-watcher", func(ctx context.Context) {
+		// We can't use a Closer since the receiver will be blocking inside of a task.
+		<-stopper.ShouldQuiesce()
+		receiver.Close()
+	}); err != nil {
+		return nil, err
+	}
 	server, err := newMockSideTransportGRPCServerWithOpts(stopper, receiver)
 	if err != nil {
 		return nil, err

--- a/pkg/rpc/context.go
+++ b/pkg/rpc/context.go
@@ -121,7 +121,7 @@ func WithInterceptor(f func(fullMethod string) error) ServerOption {
 // NewServer sets up an RPC server. Depending on the ServerOptions, the Server
 // either expects incoming connections from KV nodes, or from tenant SQL
 // servers.
-func NewServer(ctx *Context, opts ...ServerOption) *grpc.Server {
+func NewServer(rpcCtx *Context, opts ...ServerOption) *grpc.Server {
 	var o serverOpts
 	for _, f := range opts {
 		f(&o)
@@ -147,10 +147,10 @@ func NewServer(ctx *Context, opts ...ServerOption) *grpc.Server {
 		grpc.KeepaliveParams(serverKeepalive),
 		grpc.KeepaliveEnforcementPolicy(serverEnforcement),
 		// A stats handler to measure server network stats.
-		grpc.StatsHandler(&ctx.stats),
+		grpc.StatsHandler(&rpcCtx.stats),
 	}
-	if !ctx.Config.Insecure {
-		tlsConfig, err := ctx.GetServerTLSConfig()
+	if !rpcCtx.Config.Insecure {
+		tlsConfig, err := rpcCtx.GetServerTLSConfig()
 		if err != nil {
 			panic(err)
 		}
@@ -162,10 +162,10 @@ func NewServer(ctx *Context, opts ...ServerOption) *grpc.Server {
 	var unaryInterceptor []grpc.UnaryServerInterceptor
 	var streamInterceptor []grpc.StreamServerInterceptor
 
-	if !ctx.Config.Insecure {
+	if !rpcCtx.Config.Insecure {
 		a := kvAuth{
 			tenant: tenantAuthorizer{
-				tenantID: ctx.tenID,
+				tenantID: rpcCtx.tenID,
 			},
 		}
 
@@ -193,7 +193,7 @@ func NewServer(ctx *Context, opts ...ServerOption) *grpc.Server {
 		})
 	}
 
-	if tracer := ctx.AmbientCtx.Tracer; tracer != nil {
+	if tracer := rpcCtx.AmbientCtx.Tracer; tracer != nil {
 		unaryInterceptor = append(unaryInterceptor, tracing.ServerInterceptor(tracer))
 		streamInterceptor = append(streamInterceptor, tracing.StreamServerInterceptor(tracer))
 	}
@@ -202,7 +202,7 @@ func NewServer(ctx *Context, opts ...ServerOption) *grpc.Server {
 	grpcOpts = append(grpcOpts, grpc.ChainStreamInterceptor(streamInterceptor...))
 
 	s := grpc.NewServer(grpcOpts...)
-	RegisterHeartbeatServer(s, ctx.NewHeartbeatService())
+	RegisterHeartbeatServer(s, rpcCtx.NewHeartbeatService())
 	return s
 }
 
@@ -461,35 +461,35 @@ func NewContext(opts ContextOptions) *Context {
 }
 
 // ClusterName retrieves the configured cluster name.
-func (ctx *Context) ClusterName() string {
-	if ctx == nil {
+func (rpcCtx *Context) ClusterName() string {
+	if rpcCtx == nil {
 		// This is used in tests.
 		return "<MISSING RPC CONTEXT>"
 	}
-	return ctx.Config.ClusterName
+	return rpcCtx.Config.ClusterName
 }
 
 // GetStatsMap returns a map of network statistics maintained by the
 // internal stats handler. The map is from the remote network address
 // (in string form) to an rpc.Stats object.
-func (ctx *Context) GetStatsMap() *syncmap.Map {
-	return &ctx.stats.stats
+func (rpcCtx *Context) GetStatsMap() *syncmap.Map {
+	return &rpcCtx.stats.stats
 }
 
 // Metrics returns the Context's Metrics struct.
-func (ctx *Context) Metrics() *Metrics {
-	return &ctx.metrics
+func (rpcCtx *Context) Metrics() *Metrics {
+	return &rpcCtx.metrics
 }
 
 // GetLocalInternalClientForAddr returns the context's internal batch client
 // for target, if it exists.
 // Note: the node ID ought to be retyped, see
 // https://github.com/cockroachdb/cockroach/pull/73309
-func (ctx *Context) GetLocalInternalClientForAddr(
+func (rpcCtx *Context) GetLocalInternalClientForAddr(
 	target string, nodeID roachpb.NodeID,
 ) roachpb.InternalClient {
-	if target == ctx.Config.AdvertiseAddr && nodeID == ctx.NodeID.Get() {
-		return ctx.localInternalClient
+	if target == rpcCtx.Config.AdvertiseAddr && nodeID == rpcCtx.NodeID.Get() {
+		return rpcCtx.localInternalClient
 	}
 	return nil
 }
@@ -701,21 +701,21 @@ func IsLocal(iface roachpb.InternalClient) bool {
 }
 
 // SetLocalInternalServer sets the context's local internal batch server.
-func (ctx *Context) SetLocalInternalServer(internalServer roachpb.InternalServer) {
-	ctx.localInternalClient = internalClientAdapter{internalServer}
+func (rpcCtx *Context) SetLocalInternalServer(internalServer roachpb.InternalServer) {
+	rpcCtx.localInternalClient = internalClientAdapter{internalServer}
 }
 
 // removeConn removes the given connection from the pool. The supplied connKeys
 // must represent *all* the keys under among which the connection was shared.
-func (ctx *Context) removeConn(conn *Connection, keys ...connKey) {
+func (rpcCtx *Context) removeConn(conn *Connection, keys ...connKey) {
 	for _, key := range keys {
-		ctx.conns.Delete(key)
+		rpcCtx.conns.Delete(key)
 	}
-	log.Health.Infof(ctx.masterCtx, "closing %+v", keys)
+	log.Health.Infof(rpcCtx.masterCtx, "closing %+v", keys)
 	if grpcConn := conn.grpcConn; grpcConn != nil {
 		err := grpcConn.Close() // nolint:grpcconnclose
 		if err != nil && !grpcutil.IsClosedConnection(err) {
-			log.Health.Warningf(ctx.masterCtx, "failed to close client connection: %v", err)
+			log.Health.Warningf(rpcCtx.masterCtx, "failed to close client connection: %v", err)
 		}
 	}
 }
@@ -724,12 +724,14 @@ func (ctx *Context) removeConn(conn *Connection, keys ...connKey) {
 // class to the given node that succeeded on its most recent heartbeat.
 // Note: the node ID ought to be retyped, see
 // https://github.com/cockroachdb/cockroach/pull/73309
-func (ctx *Context) ConnHealth(target string, nodeID roachpb.NodeID, class ConnectionClass) error {
+func (rpcCtx *Context) ConnHealth(
+	target string, nodeID roachpb.NodeID, class ConnectionClass,
+) error {
 	// The local client is always considered healthy.
-	if ctx.GetLocalInternalClientForAddr(target, nodeID) != nil {
+	if rpcCtx.GetLocalInternalClientForAddr(target, nodeID) != nil {
 		return nil
 	}
-	if value, ok := ctx.conns.Load(connKey{target, nodeID, class}); ok {
+	if value, ok := rpcCtx.conns.Load(connKey{target, nodeID, class}); ok {
 		return value.(*Connection).Health()
 	}
 	return ErrNoConnection
@@ -743,25 +745,25 @@ func (ctx *Context) ConnHealth(target string, nodeID roachpb.NodeID, class Conne
 // particular, performance tuning options are omitted. Decompression is
 // necessarily included to support compression-enabled servers, and compression
 // is included for symmetry. These choices are admittedly subjective.
-func (ctx *Context) GRPCDialOptions() ([]grpc.DialOption, error) {
-	return ctx.grpcDialOptions("", DefaultClass)
+func (rpcCtx *Context) GRPCDialOptions() ([]grpc.DialOption, error) {
+	return rpcCtx.grpcDialOptions("", DefaultClass)
 }
 
 // grpcDialOptions extends GRPCDialOptions to support a connection class for use
 // with TestingKnobs.
-func (ctx *Context) grpcDialOptions(
+func (rpcCtx *Context) grpcDialOptions(
 	target string, class ConnectionClass,
 ) ([]grpc.DialOption, error) {
 	var dialOpts []grpc.DialOption
-	if ctx.Config.Insecure {
+	if rpcCtx.Config.Insecure {
 		dialOpts = append(dialOpts, grpc.WithInsecure())
 	} else {
 		var tlsConfig *tls.Config
 		var err error
-		if ctx.tenID == roachpb.SystemTenantID {
-			tlsConfig, err = ctx.GetClientTLSConfig()
+		if rpcCtx.tenID == roachpb.SystemTenantID {
+			tlsConfig, err = rpcCtx.GetClientTLSConfig()
 		} else {
-			tlsConfig, err = ctx.GetTenantTLSConfig()
+			tlsConfig, err = rpcCtx.GetTenantTLSConfig()
 		}
 		if err != nil {
 			return nil, err
@@ -781,7 +783,7 @@ func (ctx *Context) grpcDialOptions(
 
 	// Compression is enabled separately from decompression to allow staged
 	// rollout.
-	if ctx.rpcCompression {
+	if rpcCtx.rpcCompression {
 		dialOpts = append(dialOpts, grpc.WithDefaultCallOptions(grpc.UseCompressor((snappyCompressor{}).Name())))
 	}
 
@@ -796,7 +798,7 @@ func (ctx *Context) grpcDialOptions(
 	var unaryInterceptors []grpc.UnaryClientInterceptor
 	var streamInterceptors []grpc.StreamClientInterceptor
 
-	if tracer := ctx.AmbientCtx.Tracer; tracer != nil {
+	if tracer := rpcCtx.AmbientCtx.Tracer; tracer != nil {
 		// TODO(tbg): re-write all of this for our tracer.
 
 		// We use a decorator to set the "node" tag. All other spans get the
@@ -809,34 +811,34 @@ func (ctx *Context) grpcDialOptions(
 		// is in setupSpanForIncomingRPC().
 		//
 		tagger := func(span *tracing.Span) {
-			span.SetTag("node", attribute.IntValue(int(ctx.NodeID.Get())))
+			span.SetTag("node", attribute.IntValue(int(rpcCtx.NodeID.Get())))
 		}
 		unaryInterceptors = append(unaryInterceptors,
 			tracing.ClientInterceptor(tracer, tagger))
 		streamInterceptors = append(streamInterceptors,
 			tracing.StreamClientInterceptor(tracer, tagger))
 	}
-	if ctx.Knobs.UnaryClientInterceptor != nil {
-		testingUnaryInterceptor := ctx.Knobs.UnaryClientInterceptor(target, class)
+	if rpcCtx.Knobs.UnaryClientInterceptor != nil {
+		testingUnaryInterceptor := rpcCtx.Knobs.UnaryClientInterceptor(target, class)
 		if testingUnaryInterceptor != nil {
 			unaryInterceptors = append(unaryInterceptors, testingUnaryInterceptor)
 		}
 	}
-	if ctx.Knobs.StreamClientInterceptor != nil {
-		testingStreamInterceptor := ctx.Knobs.StreamClientInterceptor(target, class)
+	if rpcCtx.Knobs.StreamClientInterceptor != nil {
+		testingStreamInterceptor := rpcCtx.Knobs.StreamClientInterceptor(target, class)
 		if testingStreamInterceptor != nil {
 			streamInterceptors = append(streamInterceptors, testingStreamInterceptor)
 		}
 	}
-	if ctx.Knobs.ArtificialLatencyMap != nil {
+	if rpcCtx.Knobs.ArtificialLatencyMap != nil {
 		dialerFunc := func(ctx context.Context, target string) (net.Conn, error) {
 			dialer := net.Dialer{
 				LocalAddr: sourceAddr,
 			}
 			return dialer.DialContext(ctx, "tcp", target)
 		}
-		latency := ctx.Knobs.ArtificialLatencyMap[target]
-		log.VEventf(ctx.masterCtx, 1, "connecting to node %s with simulated latency %dms", target, latency)
+		latency := rpcCtx.Knobs.ArtificialLatencyMap[target]
+		log.VEventf(rpcCtx.masterCtx, 1, "connecting to node %s with simulated latency %dms", target, latency)
 		dialer := artificialLatencyDialer{
 			dialerFunc: dialerFunc,
 			latencyMS:  latency,
@@ -1022,20 +1024,20 @@ type delayingHeader struct {
 // the returned channel is closed when a reconnection is attempted.
 // This method implies a DefaultClass ConnectionClass for the returned
 // ClientConn.
-func (ctx *Context) GRPCDialRaw(target string) (*grpc.ClientConn, <-chan struct{}, error) {
-	return ctx.grpcDialRaw(target, 0, DefaultClass)
+func (rpcCtx *Context) GRPCDialRaw(target string) (*grpc.ClientConn, <-chan struct{}, error) {
+	return rpcCtx.grpcDialRaw(target, 0, DefaultClass)
 }
 
-func (ctx *Context) grpcDialRaw(
+func (rpcCtx *Context) grpcDialRaw(
 	target string, remoteNodeID roachpb.NodeID, class ConnectionClass,
 ) (*grpc.ClientConn, <-chan struct{}, error) {
-	dialOpts, err := ctx.grpcDialOptions(target, class)
+	dialOpts, err := rpcCtx.grpcDialOptions(target, class)
 	if err != nil {
 		return nil, nil, err
 	}
 
 	// Add a stats handler to measure client network stats.
-	dialOpts = append(dialOpts, grpc.WithStatsHandler(ctx.stats.newClient(target)))
+	dialOpts = append(dialOpts, grpc.WithStatsHandler(rpcCtx.stats.newClient(target)))
 
 	// Lower the MaxBackoff (which defaults to ~minutes) to something in the
 	// ~second range.
@@ -1053,9 +1055,9 @@ func (ctx *Context) grpcDialRaw(
 		redialChan: make(chan struct{}),
 	}
 	dialerFunc := dialer.dial
-	if ctx.Knobs.ArtificialLatencyMap != nil {
-		latency := ctx.Knobs.ArtificialLatencyMap[target]
-		log.VEventf(ctx.masterCtx, 1, "connecting to node %s (%d) with simulated latency %dms", target, remoteNodeID,
+	if rpcCtx.Knobs.ArtificialLatencyMap != nil {
+		latency := rpcCtx.Knobs.ArtificialLatencyMap[target]
+		log.VEventf(rpcCtx.masterCtx, 1, "connecting to node %s (%d) with simulated latency %dms", target, remoteNodeID,
 			latency)
 		dialer := artificialLatencyDialer{
 			dialerFunc: dialerFunc,
@@ -1068,10 +1070,10 @@ func (ctx *Context) grpcDialRaw(
 	// add testingDialOpts after our dialer because one of our tests
 	// uses a custom dialer (this disables the only-one-connection
 	// behavior and redialChan will never be closed).
-	dialOpts = append(dialOpts, ctx.testingDialOpts...)
+	dialOpts = append(dialOpts, rpcCtx.testingDialOpts...)
 
-	log.Health.Infof(ctx.masterCtx, "dialing n%v: %s (%v)", remoteNodeID, target, class)
-	conn, err := grpc.DialContext(ctx.masterCtx, target, dialOpts...)
+	log.Health.Infof(rpcCtx.masterCtx, "dialing n%v: %s (%v)", remoteNodeID, target, class)
+	conn, err := grpc.DialContext(rpcCtx.masterCtx, target, dialOpts...)
 	return conn, dialer.redialChan, err
 }
 
@@ -1079,8 +1081,8 @@ func (ctx *Context) grpcDialRaw(
 // node ID between client and server. This function should only be
 // used with the gossip client and CLI commands which can talk to any
 // node. This method implies a SystemClass.
-func (ctx *Context) GRPCUnvalidatedDial(target string) *Connection {
-	return ctx.grpcDialNodeInternal(target, 0, SystemClass)
+func (rpcCtx *Context) GRPCUnvalidatedDial(target string) *Connection {
+	return rpcCtx.grpcDialNodeInternal(target, 0, SystemClass)
 }
 
 // GRPCDialNode calls grpc.Dial with options appropriate for the
@@ -1090,13 +1092,13 @@ func (ctx *Context) GRPCUnvalidatedDial(target string) *Connection {
 // the remote node; this is checked during heartbeats. The caller is
 // responsible for ensuring the remote node ID is known prior to using
 // this function.
-func (ctx *Context) GRPCDialNode(
+func (rpcCtx *Context) GRPCDialNode(
 	target string, remoteNodeID roachpb.NodeID, class ConnectionClass,
 ) *Connection {
-	if remoteNodeID == 0 && !ctx.TestingAllowNamedRPCToAnonymousServer {
+	if remoteNodeID == 0 && !rpcCtx.TestingAllowNamedRPCToAnonymousServer {
 		log.Fatalf(context.TODO(), "%v", errors.AssertionFailedf("invalid node ID 0 in GRPCDialNode()"))
 	}
-	return ctx.grpcDialNodeInternal(target, remoteNodeID, class)
+	return rpcCtx.grpcDialNodeInternal(target, remoteNodeID, class)
 }
 
 // GRPCDialPod wraps GRPCDialNode and treats the `remoteInstanceID`
@@ -1106,19 +1108,19 @@ func (ctx *Context) GRPCDialNode(
 //
 // Since GRPCDialNode accepts a separate `target` and `NodeID` it
 // requires no further modification to work between pods.
-func (ctx *Context) GRPCDialPod(
+func (rpcCtx *Context) GRPCDialPod(
 	target string, remoteInstanceID base.SQLInstanceID, class ConnectionClass,
 ) *Connection {
-	return ctx.GRPCDialNode(target, roachpb.NodeID(remoteInstanceID), class)
+	return rpcCtx.GRPCDialNode(target, roachpb.NodeID(remoteInstanceID), class)
 }
 
-func (ctx *Context) grpcDialNodeInternal(
+func (rpcCtx *Context) grpcDialNodeInternal(
 	target string, remoteNodeID roachpb.NodeID, class ConnectionClass,
 ) *Connection {
 	thisConnKeys := []connKey{{target, remoteNodeID, class}}
-	value, ok := ctx.conns.Load(thisConnKeys[0])
+	value, ok := rpcCtx.conns.Load(thisConnKeys[0])
 	if !ok {
-		value, _ = ctx.conns.LoadOrStore(thisConnKeys[0], newConnectionToNodeID(ctx.Stopper, remoteNodeID))
+		value, _ = rpcCtx.conns.LoadOrStore(thisConnKeys[0], newConnectionToNodeID(rpcCtx.Stopper, remoteNodeID))
 		if remoteNodeID != 0 {
 			// If the first connection established at a target address is
 			// for a specific node ID, then we want to reuse that connection
@@ -1137,7 +1139,7 @@ func (ctx *Context) grpcDialNodeInternal(
 			// See:
 			// https://github.com/cockroachdb/cockroach/issues/37200
 			otherKey := connKey{target, 0, class}
-			if _, loaded := ctx.conns.LoadOrStore(otherKey, value); !loaded {
+			if _, loaded := rpcCtx.conns.LoadOrStore(otherKey, value); !loaded {
 				thisConnKeys = append(thisConnKeys, otherKey)
 			}
 		}
@@ -1148,22 +1150,22 @@ func (ctx *Context) grpcDialNodeInternal(
 		// Either we kick off the heartbeat loop (and clean up when it's done),
 		// or we clean up the connKey entries immediately.
 		var redialChan <-chan struct{}
-		conn.grpcConn, redialChan, conn.dialErr = ctx.grpcDialRaw(target, remoteNodeID, class)
+		conn.grpcConn, redialChan, conn.dialErr = rpcCtx.grpcDialRaw(target, remoteNodeID, class)
 		if conn.dialErr == nil {
-			if err := ctx.Stopper.RunAsyncTask(
-				ctx.masterCtx, "rpc.Context: grpc heartbeat", func(masterCtx context.Context) {
-					err := ctx.runHeartbeat(conn, target, redialChan)
+			if err := rpcCtx.Stopper.RunAsyncTask(
+				rpcCtx.masterCtx, "rpc.Context: grpc heartbeat", func(masterCtx context.Context) {
+					err := rpcCtx.runHeartbeat(conn, target, redialChan)
 					if err != nil && !grpcutil.IsClosedConnection(err) &&
 						!grpcutil.IsConnectionRejected(err) {
 						log.Health.Errorf(masterCtx, "removing connection to %s due to error: %s", target, err)
 					}
-					ctx.removeConn(conn, thisConnKeys...)
+					rpcCtx.removeConn(conn, thisConnKeys...)
 				}); err != nil {
 				conn.dialErr = err
 			}
 		}
 		if conn.dialErr != nil {
-			ctx.removeConn(conn, thisConnKeys...)
+			rpcCtx.removeConn(conn, thisConnKeys...)
 		}
 	})
 
@@ -1173,11 +1175,11 @@ func (ctx *Context) grpcDialNodeInternal(
 // NewBreaker creates a new circuit breaker properly configured for RPC
 // connections. name is used internally for logging state changes of the
 // returned breaker.
-func (ctx *Context) NewBreaker(name string) *circuit.Breaker {
-	if ctx.BreakerFactory != nil {
-		return ctx.BreakerFactory()
+func (rpcCtx *Context) NewBreaker(name string) *circuit.Breaker {
+	if rpcCtx.BreakerFactory != nil {
+		return rpcCtx.BreakerFactory()
 	}
-	return newBreaker(ctx.masterCtx, name, &ctx.breakerClock)
+	return newBreaker(rpcCtx.masterCtx, name, &rpcCtx.breakerClock)
 }
 
 // ErrNotHeartbeated is returned by ConnHealth when we have not yet performed
@@ -1188,13 +1190,13 @@ var ErrNotHeartbeated = errors.New("not yet heartbeated")
 // the node.
 var ErrNoConnection = errors.New("no connection found")
 
-func (ctx *Context) runHeartbeat(
+func (rpcCtx *Context) runHeartbeat(
 	conn *Connection, target string, redialChan <-chan struct{},
 ) (retErr error) {
-	ctx.metrics.HeartbeatLoopsStarted.Inc(1)
+	rpcCtx.metrics.HeartbeatLoopsStarted.Inc(1)
 	// setInitialHeartbeatDone is idempotent and is critical to notify Connect
 	// callers of the failure in the case where no heartbeat is ever sent.
-	state := updateHeartbeatState(&ctx.metrics, heartbeatNotRunning, heartbeatInitializing)
+	state := updateHeartbeatState(&rpcCtx.metrics, heartbeatNotRunning, heartbeatInitializing)
 	initialHeartbeatDone := false
 	setInitialHeartbeatDone := func() {
 		if !initialHeartbeatDone {
@@ -1204,12 +1206,12 @@ func (ctx *Context) runHeartbeat(
 	}
 	defer func() {
 		if retErr != nil {
-			ctx.metrics.HeartbeatLoopsExited.Inc(1)
+			rpcCtx.metrics.HeartbeatLoopsExited.Inc(1)
 		}
-		updateHeartbeatState(&ctx.metrics, state, heartbeatNotRunning)
+		updateHeartbeatState(&rpcCtx.metrics, state, heartbeatNotRunning)
 		setInitialHeartbeatDone()
 	}()
-	maxOffset := ctx.Clock.MaxOffset()
+	maxOffset := rpcCtx.Clock.MaxOffset()
 	maxOffsetNanos := maxOffset.Nanoseconds()
 
 	heartbeatClient := NewHeartbeatClient(conn.grpcConn)
@@ -1231,31 +1233,31 @@ func (ctx *Context) runHeartbeat(
 		select {
 		case <-redialChan:
 			return grpcutil.ErrCannotReuseClientConn
-		case <-ctx.Stopper.ShouldQuiesce():
+		case <-rpcCtx.Stopper.ShouldQuiesce():
 			return nil
 		case <-heartbeatTimer.C:
 			heartbeatTimer.Read = true
 		}
 
-		if err := ctx.Stopper.RunTaskWithErr(ctx.masterCtx, "rpc heartbeat", func(goCtx context.Context) error {
+		if err := rpcCtx.Stopper.RunTaskWithErr(rpcCtx.masterCtx, "rpc heartbeat", func(goCtx context.Context) error {
 			// We re-mint the PingRequest to pick up any asynchronous update to clusterID.
-			clusterID := ctx.ClusterID.Get()
+			clusterID := rpcCtx.ClusterID.Get()
 			request := &PingRequest{
-				OriginNodeID:         ctx.NodeID.Get(),
-				OriginAddr:           ctx.Config.Addr,
+				OriginNodeID:         rpcCtx.NodeID.Get(),
+				OriginAddr:           rpcCtx.Config.Addr,
 				OriginMaxOffsetNanos: maxOffsetNanos,
 				ClusterID:            &clusterID,
 				TargetNodeID:         conn.remoteNodeID,
-				ServerVersion:        ctx.Settings.Version.BinaryVersion(),
+				ServerVersion:        rpcCtx.Settings.Version.BinaryVersion(),
 			}
 
 			interceptor := func(*PingRequest) error { return nil }
-			if fn := ctx.OnOutgoingPing; fn != nil {
+			if fn := rpcCtx.OnOutgoingPing; fn != nil {
 				interceptor = fn
 			}
 
 			var response *PingResponse
-			sendTime := ctx.Clock.PhysicalTime()
+			sendTime := rpcCtx.Clock.PhysicalTime()
 			ping := func(goCtx context.Context) error {
 				// NB: We want the request to fail-fast (the default), otherwise we won't
 				// be notified of transport failures.
@@ -1268,8 +1270,8 @@ func (ctx *Context) runHeartbeat(
 				return err
 			}
 			var err error
-			if ctx.heartbeatTimeout > 0 {
-				err = contextutil.RunWithTimeout(goCtx, "rpc heartbeat", ctx.heartbeatTimeout, ping)
+			if rpcCtx.heartbeatTimeout > 0 {
+				err = contextutil.RunWithTimeout(goCtx, "rpc heartbeat", rpcCtx.heartbeatTimeout, ping)
 			} else {
 				err = ping(goCtx)
 			}
@@ -1285,9 +1287,9 @@ func (ctx *Context) runHeartbeat(
 				// new node in a cluster and mistakenly joins the wrong
 				// cluster gets a chance to see the error message on their
 				// management console.
-				if !ctx.Config.DisableClusterNameVerification && !response.DisableClusterNameVerification {
+				if !rpcCtx.Config.DisableClusterNameVerification && !response.DisableClusterNameVerification {
 					err = errors.Wrap(
-						checkClusterName(ctx.Config.ClusterName, response.ClusterName),
+						checkClusterName(rpcCtx.Config.ClusterName, response.ClusterName),
 						"cluster name check failed on ping response")
 					if err != nil {
 						returnErr = true
@@ -1297,7 +1299,7 @@ func (ctx *Context) runHeartbeat(
 
 			if err == nil {
 				err = errors.Wrap(
-					checkVersion(goCtx, ctx.Settings, response.ServerVersion),
+					checkVersion(goCtx, rpcCtx.Settings, response.ServerVersion),
 					"version compatibility check failed on ping response")
 				if err != nil {
 					returnErr = true
@@ -1306,12 +1308,12 @@ func (ctx *Context) runHeartbeat(
 
 			if err == nil {
 				everSucceeded = true
-				receiveTime := ctx.Clock.PhysicalTime()
+				receiveTime := rpcCtx.Clock.PhysicalTime()
 
 				// Only update the clock offset measurement if we actually got a
 				// successful response from the server.
 				pingDuration := receiveTime.Sub(sendTime)
-				maxOffset := ctx.Clock.MaxOffset()
+				maxOffset := rpcCtx.Clock.MaxOffset()
 				if pingDuration > maximumPingDurationMult*maxOffset {
 					request.Offset.Reset()
 				} else {
@@ -1325,9 +1327,9 @@ func (ctx *Context) runHeartbeat(
 					remoteTimeNow := timeutil.Unix(0, response.ServerTime).Add(pingDuration / 2)
 					request.Offset.Offset = remoteTimeNow.Sub(receiveTime).Nanoseconds()
 				}
-				ctx.RemoteClocks.UpdateOffset(ctx.masterCtx, target, request.Offset, pingDuration)
+				rpcCtx.RemoteClocks.UpdateOffset(rpcCtx.masterCtx, target, request.Offset, pingDuration)
 
-				if cb := ctx.HeartbeatCB; cb != nil {
+				if cb := rpcCtx.HeartbeatCB; cb != nil {
 					cb()
 				}
 			}
@@ -1336,7 +1338,7 @@ func (ctx *Context) runHeartbeat(
 				everSucceeded: everSucceeded,
 				err:           err,
 			}
-			state = updateHeartbeatState(&ctx.metrics, state, hr.state())
+			state = updateHeartbeatState(&rpcCtx.metrics, state, hr.state())
 			conn.heartbeatResult.Store(hr)
 			setInitialHeartbeatDone()
 			if returnErr {
@@ -1347,21 +1349,21 @@ func (ctx *Context) runHeartbeat(
 			return err
 		}
 
-		heartbeatTimer.Reset(ctx.Config.RPCHeartbeatInterval)
+		heartbeatTimer.Reset(rpcCtx.Config.RPCHeartbeatInterval)
 	}
 }
 
 // NewHeartbeatService returns a HeartbeatService initialized from the Context.
-func (ctx *Context) NewHeartbeatService() *HeartbeatService {
+func (rpcCtx *Context) NewHeartbeatService() *HeartbeatService {
 	return &HeartbeatService{
-		clock:                                 ctx.Clock,
-		remoteClockMonitor:                    ctx.RemoteClocks,
-		clusterName:                           ctx.ClusterName(),
-		disableClusterNameVerification:        ctx.Config.DisableClusterNameVerification,
-		clusterID:                             ctx.ClusterID,
-		nodeID:                                ctx.NodeID,
-		settings:                              ctx.Settings,
-		onHandlePing:                          ctx.OnIncomingPing,
-		testingAllowNamedRPCToAnonymousServer: ctx.TestingAllowNamedRPCToAnonymousServer,
+		clock:                                 rpcCtx.Clock,
+		remoteClockMonitor:                    rpcCtx.RemoteClocks,
+		clusterName:                           rpcCtx.ClusterName(),
+		disableClusterNameVerification:        rpcCtx.Config.DisableClusterNameVerification,
+		clusterID:                             rpcCtx.ClusterID,
+		nodeID:                                rpcCtx.NodeID,
+		settings:                              rpcCtx.Settings,
+		onHandlePing:                          rpcCtx.OnIncomingPing,
+		testingAllowNamedRPCToAnonymousServer: rpcCtx.TestingAllowNamedRPCToAnonymousServer,
 	}
 }

--- a/pkg/rpc/context.go
+++ b/pkg/rpc/context.go
@@ -166,9 +166,29 @@ func NewServer(rpcCtx *Context, opts ...ServerOption) *grpc.Server {
 	}
 
 	// These interceptors will be called in the order in which they appear, i.e.
-	// The last element will wrap the actual handler.
+	// The last element will wrap the actual handler. The first interceptor
+	// guards RPC endpoints for use after Stopper.Drain() by handling the RPC
+	// inside a stopper task.
 	var unaryInterceptor []grpc.UnaryServerInterceptor
 	var streamInterceptor []grpc.StreamServerInterceptor
+	unaryInterceptor = append(unaryInterceptor, func(
+		ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler,
+	) (interface{}, error) {
+		var resp interface{}
+		if err := rpcCtx.Stopper.RunTaskWithErr(ctx, info.FullMethod, func(ctx context.Context) error {
+			var err error
+			resp, err = handler(ctx, req)
+			return err
+		}); err != nil {
+			return nil, err
+		}
+		return resp, nil
+	})
+	streamInterceptor = append(streamInterceptor, func(srv interface{}, ss grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+		return rpcCtx.Stopper.RunTaskWithErr(ss.Context(), info.FullMethod, func(ctx context.Context) error {
+			return handler(srv, ss)
+		})
+	})
 
 	if !rpcCtx.Config.Insecure {
 		a := kvAuth{

--- a/pkg/rpc/context.go
+++ b/pkg/rpc/context.go
@@ -45,9 +45,11 @@ import (
 	"golang.org/x/sync/syncmap"
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/backoff"
+	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/credentials"
 	"google.golang.org/grpc/encoding"
 	"google.golang.org/grpc/metadata"
+	grpcstatus "google.golang.org/grpc/status"
 )
 
 func init() {
@@ -74,6 +76,12 @@ const (
 // GRPC Dialer connection timeout. 20s matches default value that is
 // suppressed when backoff config is provided.
 const minConnectionTimeout = 20 * time.Second
+
+// errDialRejected is returned from client interceptors when the server's
+// stopper is quiescing. The error is constructed to return true in
+// `grpcutil.IsConnectionRejected` which prevents infinite retry loops during
+// cluster shutdown, especially in unit testing.
+var errDialRejected = grpcstatus.Error(codes.PermissionDenied, "refusing to dial; node is quiescing")
 
 // sourceAddr is the environment-provided local address for outgoing
 // connections.
@@ -447,7 +455,7 @@ func NewContext(opts ContextOptions) *Context {
 				// conn. We need to set the error in case we win the race against the
 				// real initialization code.
 				if conn.dialErr == nil {
-					conn.dialErr = &roachpb.NodeUnavailableError{}
+					conn.dialErr = errDialRejected
 				}
 			})
 			ctx.removeConn(conn, k.(connKey))
@@ -1074,6 +1082,12 @@ func (rpcCtx *Context) grpcDialRaw(
 
 	log.Health.Infof(rpcCtx.masterCtx, "dialing n%v: %s (%v)", remoteNodeID, target, class)
 	conn, err := grpc.DialContext(rpcCtx.masterCtx, target, dialOpts...)
+	if err != nil && rpcCtx.masterCtx.Err() != nil {
+		// If the node is draining, discard the error (which is likely gRPC's version
+		// of context.Canceled) and return errDialRejected which instructs callers not
+		// to retry.
+		err = errDialRejected
+	}
 	return conn, dialer.redialChan, err
 }
 
@@ -1161,7 +1175,10 @@ func (rpcCtx *Context) grpcDialNodeInternal(
 					}
 					rpcCtx.removeConn(conn, thisConnKeys...)
 				}); err != nil {
-				conn.dialErr = err
+				// If node is draining (`err` will always equal stop.ErrUnavailable
+				// here), return special error (see its comments).
+				_ = err // ignore this error
+				conn.dialErr = errDialRejected
 			}
 		}
 		if conn.dialErr != nil {

--- a/pkg/rpc/context_test.go
+++ b/pkg/rpc/context_test.go
@@ -1844,3 +1844,60 @@ func BenchmarkGRPCDial(b *testing.B) {
 		}
 	})
 }
+
+func TestRejectDialOnQuiesce(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	ctx := context.Background()
+
+	clusID := uuid.MakeV4()
+	clock := hlc.NewClock(hlc.UnixNano, 250*time.Millisecond)
+	// Run vanilla server.
+	const serverNodeID = 1
+	var addr string
+	{
+		srvStopper := stop.NewStopper()
+		defer srvStopper.Stop(ctx)
+		serverCtx := newTestContext(clusID, clock, srvStopper)
+		serverCtx.NodeID.Set(ctx, serverNodeID)
+		s := NewServer(serverCtx)
+		ln, err := netutil.ListenAndServeGRPC(srvStopper, s, util.TestAddr)
+		require.NoError(t, err)
+		addr = ln.Addr().String()
+	}
+
+	// Set up client.
+	stopper := stop.NewStopper()
+	defer stopper.Stop(ctx)
+	rpcCtx := newTestContext(clusID, clock, stopper)
+
+	// Set up one connection before quiesce to exercise the path in which
+	// the connection already exists and is healthy when the node shuts
+	// down.
+	conn, err := rpcCtx.GRPCDialNode(addr, serverNodeID, SystemClass).Connect(ctx)
+	require.NoError(t, err)
+	// Check that we can reach the server (though we don't bother to also register
+	// an endpoint on it; getting codes.Unimplemented back is proof enough).
+	err = conn.Invoke(ctx, "/Does/Not/Exist", &PingRequest{}, &PingResponse{})
+	require.Error(t, err)
+	require.Equal(t, codes.Unimplemented, status.Code(err))
+
+	// Now quiesce.
+	stopper.Quiesce(ctx)
+
+	// First, we shouldn't be able to dial again, even though we already have a
+	// connection.
+	_, err = rpcCtx.GRPCDialNode(addr, serverNodeID, SystemClass).Connect(ctx)
+	require.ErrorIs(t, err, errDialRejected)
+	require.True(t, grpcutil.IsConnectionRejected(err))
+	require.True(t, grpcutil.IsAuthError(err))
+
+	// If we use the existing connection, we'll get an error. It won't be the
+	// ideal one - it's an opaque grpc version of `context.Canceled` but it's
+	// the best we can hope for, since this is entirely within gRPC code.
+	// This is good enough to prevent hangs since callers won't be using
+	// this connection going forward; they need to call .Connect(ctx) again
+	// which gives the permanent error PermissionDenied.
+	err = conn.Invoke(ctx, "/Does/Not/Exist", &PingRequest{}, &PingResponse{})
+	require.Error(t, err)
+	require.Equal(t, codes.Canceled, status.Code(err))
+}

--- a/pkg/rpc/context_test.go
+++ b/pkg/rpc/context_test.go
@@ -62,19 +62,19 @@ import (
 // instead which automates the address resolution.
 //
 // TODO(knz): remove this altogether. Use the dialer in all cases.
-func (ctx *Context) TestingConnHealth(target string, nodeID roachpb.NodeID) error {
-	if ctx.GetLocalInternalClientForAddr(target, nodeID) != nil {
+func (rpcCtx *Context) TestingConnHealth(target string, nodeID roachpb.NodeID) error {
+	if rpcCtx.GetLocalInternalClientForAddr(target, nodeID) != nil {
 		// The local server is always considered healthy.
 		return nil
 	}
-	conn := ctx.GRPCDialNode(target, nodeID, DefaultClass)
+	conn := rpcCtx.GRPCDialNode(target, nodeID, DefaultClass)
 	return conn.Health()
 }
 
 // AddTestingDialOpts adds extra dialing options to the rpc Context. This should
 // be done before GRPCDial is called.
-func (ctx *Context) AddTestingDialOpts(opts ...grpc.DialOption) {
-	ctx.testingDialOpts = append(ctx.testingDialOpts, opts...)
+func (rpcCtx *Context) AddTestingDialOpts(opts ...grpc.DialOption) {
+	rpcCtx.testingDialOpts = append(rpcCtx.testingDialOpts, opts...)
 }
 
 func newTestServer(t testing.TB, ctx *Context, extraOpts ...grpc.ServerOption) *grpc.Server {

--- a/pkg/server/node.go
+++ b/pkg/server/node.go
@@ -1355,6 +1355,8 @@ func (n *Node) GossipSubscription(
 			}
 		case <-ctxDone:
 			return ctx.Err()
+		case <-n.stopper.ShouldQuiesce():
+			return stop.ErrUnavailable
 		}
 	}
 }

--- a/pkg/util/grpcutil/grpc_util.go
+++ b/pkg/util/grpcutil/grpc_util.go
@@ -110,6 +110,13 @@ func IsConnectionRejected(err error) bool {
 
 // IsAuthError returns true if err's Cause is an error produced by
 // gRPC due to an authentication or authorization error for the operation.
+// AuthErrors should generally be considered non-retriable. They indicate
+// that the operation would not succeed even if directed at another node
+// in the cluster.
+//
+// As a special case, an AuthError (PermissionDenied) is returned on outbound
+// dialing when the source node is in the process of terminating (see
+// rpc.errDialRejected).
 func IsAuthError(err error) bool {
 	if s, ok := status.FromError(errors.UnwrapAll(err)); ok {
 		switch s.Code() {

